### ISSUE TITLE
Add support for SimpleConsumer-like access

### ISSUE THREAD

**Issue by ewencp** [confluentinc/kafka-rest#26]
Sometimes apps might not want to be part of a consumer group, or might have trivial consumption patterns such as consuming from a single partition. We can probably expose this pretty cleanly with URLs like GET /topics/foo/partitions/0/messages?offset=100[&count=100], although specifying per-partition offsets at the topic level may get messy.

**Comment by HenriJ** [comment]
Hi @ewencp,
I started working on this issue because I wanted to learn more about Kafka's SimpleConsumer.
Do you accept PR ?
If yes, I have something working and following your URL specification. I still have some work to make it cleaner and to add some tests.
I also have some doubts for what the values should be for the SimpleConsummer's Buffer size and Fetch Size. Right now it's hardcoded as "1024 * 1024". Any ideas :) ?

**Comment by ewencp** [comment]
@HenriJ We absolutely accepts PRs, please submit it!

One thing with this patch is that we'll want to be careful about compatibility with the new consumer -- ultimately we'd like to swap out the current implementation to use the new consumer that will be available in Kafka 0.9 for all consumer code. This means we need to think carefully about whether we can ensure a backwards-compatible implementation of the simple consumer APIs we add. I don't think this will be a problem with the new consumer's `seek()` method, but we need to think it through.

For configuration, I'll take a look at your PR, but probably it makes sense to leave it at the default for now (until we have a good reason for changing it), but expose control over it to the person managing the kafka-rest instance. This lets them override it if they really need to, but doesn't have the same set of problems as exposing that setting directly in the request.

**Comment by HenriJ** [comment]
Great if you accept PR :)
I am working on the tests now, but I already commited the feature itself on the fork https://github.com/HenriJ/kafka-rest
Don't hesitate to tell me if you feel it should have been done differently or if you see things that could be simplified / written better !

**Comment by ewencp** [comment]
Awesome, this looks like a great start. I spotted a few things with a quick review. Here are some initial notes:

1. Some defaults were changed that look like they're for your local setup.
2. There's some code SimpleConsumerFetcher that should probably be refactored to MetadataObserver, specifically the broker info since some form of that should be exposed in the API as well. One reason for trying to keep this all in MetadataObserver is that we may eventually try to replace that implementation with something that proactively maintains a copy of the metadata by watching ZK rather than always looking it up on demand.
3. I'm wondering if some of the metadata lookups could be made more efficient by not looking up data for all partitions?
4. Would we be better off with a pool of SimpleConsumers rather than generating one per request? It does introduce a bit of complexity since we need some shared state and we probably want to support more than one instance per broker, but if a consumer is making multiple calls to the API, it'd be nice to avoid the connection setup cost (let alone the log noise that it tends to generate in Kafka).
5. Needs some tests! It can be tricky to figure out how best to write these since we have both unit and integration tests. Let me know if you need guidance here.

But the basic setup looks good and this is a great feature to add! Whenever you're ready, feel free to send a pull request and I can do a more thorough review. We expect PRs to be works in progress, so don't worry if it's not entirely complete yet.

**Comment by HenriJ** [comment]
Yes, I asked myself the same question, whether I should use a SimpleConsumers pool. That will be interesting to implement !
What strategy should be used to remove a SimpleConsumer from the pool ? Should a consumer simply be removed when it has not been used since a certain amount of time (in that case do we use a TimerTask ?)
You say we want to support more than one instance per broker : should we create a new SimpleConsumer as soon as we have two concurrent requests to the same broker ?

**Comment by HenriJ** [comment]
I'm reading some integration tests to understand how it works. If I'm not mistaken, they need a local Kafka instance running in order to work. Would it be a good idea to include a Kafka instance in the project and run it with the test (on a random port), hence making the tests easier to run ?

**Comment by ewencp** [comment]
The policy for managing the pool is the really tricky part. I think the easiest solution to start with is to allow a pool of up to N SimpleConsumer instances per broker, where N is configurable. We can allocate them on-demand as needed, i.e. if we are still under the limit and none are available. I think the pool should probably be the last item to worry about thought -- it would be isolated to the one class and easy to add.

For the integration tests, they run Kafka for you if you use the ClusterTestHarness. They'll actually pull up all the services you need (ZK, Kafka, schema-registry), which is why they can be relatively slow. This made me realize I wanted to fix that, so I just filed #62 to try to address how slow they are.

**Comment by HenriJ** [comment]
It is taking shape : I got the integration tests working (SimpleConsumerAvroTest & SimpleConsumerBinaryTest)

I just have left the unit tests, but it appears that EmbeddedServerTestHarness.request() doesn't allow me to create a request with query parameters (in this case ?offset=X&count==Y). I am missing something or should I checkout your rest-utils-test project to add query param support ?

**Comment by ewencp** [comment]
That method returns the Invocation.Builder so you should be able to do pretty much anything you could do with the raw API -- it's just a small wrapper for a very common construction used in our tests. However, it doesn't look like the Invocation.Builder API supports specifying query params separately; I guess you're just supposed to add them to the target. So something like `this.request("/path/to/resource?offset=X&count=Y", mediatype)` should work.

**Comment by HenriJ** [comment]
When I do that, it converts the "?" into the url-encoded %3F, and the requests returns therefore a 404.

The Builder doesn't seem to support query params (directly at least), but the WebTarget API does : it's what I used for the integration tests. EmbeddedServerTestHarness.request() does create a WebTarget, but returns the Builder:
Builder builder = this.getJerseyTest().target(target).request(); // target() returns a WebTarget

The ClusterTestHarness in the IT exposes directly a request method, that I was able to overload to accept query params.

**Comment by ewencp** [comment]
Ah, I see. In fact there are at least a couple of examples of other places in the unit tests where I've used those APIs directly instead of using the helper method, e.g. https://github.com/confluentinc/kafka-rest/blob/master/src/test/java/io/confluent/kafkarest/unit/ConsumerResourceBinaryTest.java#L82

Using those APIs directly is fine, and it's what I'd recommend you do for now. We should create shorthand wrappers if we see the same form being used repeatedly, but we don't have that many query param calls yet. A wrapper would be nice, but we'd also have to figure out an API for it that supports multiple query params, is easy to use, and doesn't end up just replicating the WebTarget or Invocation.Builder APIs.

**Comment by HenriJ** [comment]
Hi @ewencp 
I added the UT for the new endpoints. I have a last problem : my TopicsResourceBinaryConsumeTest fails, and I don't know why.

It fails, when it runs the request with the mediatype.header=null
The application answers with "vnd.kafka.binary.v1+json", but the expected mediatype is "application/vnd.kafka.v1+json".

Any idea ?

**Comment by ewencp** [comment]
@HenriJ Yes, this is a special case. When you leave the Accept header out, what you want is to use the most specific value you can. For almost every API call so far, that has been application/vnd.kafka.v1+json. However, for API calls that return embedded data, it should be more specific than that.

It's hard to capture this clearly in the constants we have in TestUtils unfortunately. You can see how we have to use this same exception when reading data in other tests: https://github.com/confluentinc/kafka-rest/blob/master/src/test/java/io/confluent/kafkarest/unit/ConsumerResourceBinaryTest.java#L88 If you can think of a clean way to not duplicate those lists but get a separate list with just that case overridden, I would be happy to see that :) If not, the minor tweak should at least let you fix the test.

Another issue I noticed -- you used AutoCloseable which is JDK 7+. Since we're still supporting JDK6 I'd like to avoid that for now. You can just remove the interface and the @Overrides annotation and it should be fine.

Do you want to generate a pull request for this? It's a lot easier to comment on progress and track outstanding issues that need to be fixed if we have a dedicated space and can annotate the code directly. Besides, it seems like this will be nearing completion soon anyway.

**Comment by ewencp** [comment]
Oh, also with regards to content types: I think you should add

```
Versions.KAFKA_V1_JSON_BINARY_WEIGHTED_LOW, Versions.KAFKA_V1_JSON_AVRO_WEIGHTED_LOW,
```

to the `@Produces` annotations on the resources you've added methods to. This makes those resources a bit more liberal in allowing users to request content types that use embedded data even though for most API calls they shouldn't. While I'd like to be strict about it, it's probably better for the user if they can just use one content type throughout. Since we actually have methods producing those formats, we can at least allow them for other methods on the resource and just discourage them by using the lower weights that those constants attach.

**Comment by ewencp** [comment]
Fixed by #66
